src/cape_ros_node.cpp

```diff
@@ -0,0 +1,337 @@
+  auto image_depth16 = reinterpret_cast<const uint16_t*>(image_ptr->image.data);
+  float scaled_depth(0.);
+
+  if (!cape_extractor_ || height_ != image->height || width_ != image->width)
+  {
+    height_ = image->height;
+    width_ = image->width;
+    reset();
+  }
+
+  auto iter_mat = depth_mat_.begin<float>();
+
+  // Go through the depth iamge and scale each pixel
+  for (int y = 0; y < height_; ++y)
+  {
+    for (int x = 0; x < width_; ++x)
+    {
+      scaled_depth = static_cast<float>(*image_depth16) * depth_scale_;
+
+      // Check if the point is valid
+      if (scaled_depth <= 0.f)
+      {
+        *iter_mat = 0.;
+      }
+      else
+        *iter_mat = scaled_depth;
+
+      ++image_depth16;
+      ++iter_mat;
+    }
+  }
+
+  X_ = X_pre_.mul(depth_mat_);
+  Y_ = Y_pre_.mul(depth_mat_);
+  cloud_array_.setZero();
+  seg_output_.setTo(0);
+  plane_params_.clear();
+  cylinder_params_.clear();
+
+  simpleProjectPointCloud(X_, Y_, depth_mat_, cloud_array_, depth_cutoff_);
+
+  organizePointCloudByCell(cloud_array_, cloud_array_organized_, cell_map_);
+
+  cape_extractor_->process(cloud_array_organized_, seg_output_, plane_params_, cylinder_params_);
+
+  planes_pub_.publish(generateMessage(image->header));
+
+  if (intensity_image_ptr_ && (image_overlay_pub_.getNumSubscribers() > 0))
+  {
+    overlaySegmentsOnImage(intensity_image_ptr_->image, seg_output_);
+    image_overlay_pub_.publish(cv_bridge::CvImage(image->header, "rgb8", overlay_image_).toImageMsg());
+  }
+}
+
+void CapeRosNode::reset()
+{
+  cape_extractor_ = std::make_shared<Cape>(height_, width_, patch_size_, patch_size_, cylinder_detection_,
+                                           cos_angle_max_, max_merge_dist_);
+  depth_mat_.create(height_, width_, CV_32F);
+  X_.create(height_, width_, CV_32F);
+  Y_.create(height_, width_, CV_32F);
+  X_pre_.create(height_, width_, CV_32F);
+  Y_pre_.create(height_, width_, CV_32F);
+
+  cloud_array_.resize(width_ * height_, 3);
+  cloud_array_organized_.resize(width_ * height_, 3);
+
+  for (auto r = 0; r < height_; ++r)
+  {
+    for (auto c = 0; c < width_; ++c)
+    {
+      X_pre_.at<float>(r, c) = static_cast<float>((c - cx_) / fx_);
+      Y_pre_.at<float>(r, c) = static_cast<float>((r - cy_) / fy_);
+    }
+  }
+
+  // Pre-computations for maping an image point cloud to a cache-friendly array where cell's local point clouds are
+  // contiguous
+  cell_map_.create(height_, width_, CV_32S);
+
+  const auto nr_horizontal_cells = width_ / patch_size_;
+  ///@todo(afakih) join this with the loop above
+  for (int r = 0; r < height_; ++r)
+  {
+    int cell_r = r / patch_size_;
+    int local_r = r % patch_size_;
+    for (int c = 0; c < width_; ++c)
+    {
+      int cell_c = c / patch_size_;
+      int local_c = c % patch_size_;
+      cell_map_.at<int>(r, c) =
+          (cell_r * nr_horizontal_cells + cell_c) * patch_size_ * patch_size_ + local_r * patch_size_ + local_c;
```

Comment:
Is `cell_r ` `cell_c `, `local_c` 16-bit signed integer? But  `cell_map` is 32-bit signed integer?